src/sentry/integrations/source_code_management/repository.py

```diff
@@ -19,7 +19,28 @@
 REPOSITORY_INTEGRATION_GET_FILE_METRIC = "repository_integration.get_file.{result}"
 
 
-class RepositoryIntegration(IntegrationInstallation, ABC):
+class BaseRepositoryIntegration(ABC):
+    @abstractmethod
+    def get_repositories(self, query: str | None = None) -> Sequence[dict[str, Any]]:
+        """
+        Get a list of available repositories for an installation
+
+        >>> def get_repositories(self):
+        >>>     return self.get_client().get_repositories()
+
+        return [{
+            'name': display_name,
+            'identifier': external_repo_id,
+        }]
+
+        The shape of the `identifier` should match the data
+        returned by the integration's
+        IntegrationRepositoryProvider.repository_external_slug()
+        """
+        raise NotImplementedError
```

Comment:
Do you anticipate us moving more of the RepositoryIntegration methods into this new base class? Seems silly to have a new base class with a single new method, but I get why this was needed to handle the diamond inheritance issue.

Reply:
I don't really mind having the common base class between `SourceCodeIssueIntegration` and `RepositoryIntegration` even if it is just this.

I don't know if we're relying on the polymorphism anywhere -- i.e., code that calls `get_repositories` an object without knowing whether it's a `SourceCodeIssueIntegration` or a`RepositoryIntegration`. I doubt it. By that logic, you could argue that it would be fine for `SourceCodeIssueIntegration` and `RepositoryIntegration` each to have their own, separate `get_repositories` method. IMHO it comes down to communicating intent. If each class declares its own `get_repositories` method, that communicates that they're semantically unrelated, and that it should be fine to change the signature or return type on one of them without affecting the other. But if both inherit the abstract method from the same base class, it communicates to humans that `get_repositories` does the same thing with the same rules on both classes, even if the software doesn't care.

Considering the amount of information communicated in the docstring, I fall heavily on the latter side in this case. (Put more simply: we don't want to copy-and-paste that long docstring, do we?) In fact, the implicitly typed dict as the return value is begging to be refactored as it is, so all the better if we refactor it from a single base class.

Reply:
>IMHO it comes down to communicating intent. If each class declares its own get_repositories method, that communicates that they're semantically unrelated, and that it should be fine to change the signature or return type on one of them without affecting the other.

I agree, and this is the code smelly part of this implementation, which makes me wonder if we're overloading this method for multiple distinct use cases; although I reckon the implementation will be identical for now. Not a big deal, but ideally we'd reduce the number of classes we unnecessarily inherit from where we can.